### PR TITLE
introduction to avl trees

### PR DESCRIPTION
**Pull request by bloodprashure**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      

#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tool:** [Grammarly](https://grammarly.com)
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

This fixes issue #531

### CODE REVIEW

**Comment by hectorkambow** [comment]
plagiarism check done

**Comment by hectorkambow** [comment]
@bloodprashure  Lets make sure we are going down the checklist provide on the PR to make sure your PR is 100% ready to be reviewed. 

🚀

**Comment by bloodprashure** [comment]
> @bloodprashure Lets make sure we are going down the checklist provide on the PR to make sure your PR is 100% ready to be reviewed.
> 
> 🚀

Sure thing @hectorkambow. Do I need to do anything? This is my first article, so 😅

**Comment by cyberShaw** [comment]
Great topic for a first article, @bloodprashure!

We have your article reviewed in two phases. We'll initially have the preliminary review where we check for grammatical errors, readability enhancements and proper formatting. The second phase would be a tech review. In this phase, your article would be checked for technical accuracy (including code snippets).

**Comment by bloodprashure** [comment]
> Great topic for a first article, @bloodprashure!
> 
> Your article would be reviewed in two phases. We would initially have the preliminary review where we check for grammatical errors, readability enhancements and proper formatting. The second phase would be a tech review. In this phase, your article would be checked for technical accuracy (including code snippets).

Thanks @cyberShaw! Appreciate it!

**Comment by hectorkambow** [comment]
plagiarism check done

File: articles/introduction-to-avl-trees/index.md [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+
+Let's see how the insertion of a node into the following tree would affect its structure.
+
+Take the example of the following AVL Tree. 
+
+<center> <img src="/engineering-education/introduction-to-avl-trees/avl_insertion_before.jpg" alt="Before insertion"> </center>
+
+On inserting a node (4) in the tree, the structure would change to -
+
+<center> <img src="/engineering-education/introduction-to-avl-trees/tree_after_insertion.jpg" alt="After Insertion"> </center>
```
Comment (cyberShaw) [inline articles/introduction-to-avl-trees/index.md:39]: Please include image source attributions for all the images taken from an external source.

Example:

```md
*Figure: Client Side Rendering -- [Source](https://medium.com/walmartglobaltech/the-benefits-of-server-side-rendering-over-client-side-rendering-5d07ff2cefe8)*
```

File: articles/introduction-to-avl-trees/index.md [hunk 2]
```diff
@@ -0,0 +1,548 @@
+### Introduction to AVL trees
+Binary search trees are one of the most efficient data structures with their O(n) time complexity while performing operations. But in this fast paced world, especially in the realm of computers and processing, the one parameter that everything is scrutinized by is speed. The AVL tree, named after its two inventors, G.M. Abelson-Velvety and E.M. Landis, who published it in their 1962 paper "An Algorithm For The Organization Of Information" has anchored its position as a need-to-understand data structure owing to its performance increase from a regular BST.
+
```
Comment (cyberShaw) [inline articles/introduction-to-avl-trees/index.md:3]: Your article heavily relies on the reader's knowledge of Binary Trees and Binary Search Trees. Therefore, please add a prerequisites section in your article that highlights the requirements to read this article. 

Also, add a couple of links that would help the reader get to speed with the concepts related to BSTs.

File: articles/introduction-to-avl-trees/index.md [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+        - Right-right rotation: Where 'c' is the right child of 'd' and 'b' is the right child of 'c' 
+        - Left-right rotation: Where 'c' is the left child of 'd' and 'b' is the right child of 'c' 
+        - Right-left rotation: Where 'c' is the right child of 'd' and 'b' is the left child of 'c' 
+    4. After the rotations are performed according to the situation, the node is successfully inserted and the tree is balanced.
+  
+Left-left Rotation
+
+<center> <img src="/engineering-education/introduction-to-avl-trees/leftleft_rotation.jpg" alt="Left-Left Rotation"> </center>
+
+Left-right rotation
```
Comment (cyberShaw) [inline articles/introduction-to-avl-trees/index.md:65]: > Left-right rotation 

For consistency, capitalize all the image headings.
Example, 

**Left-right Rotation**

File: articles/introduction-to-avl-trees/index.md [hunk 4]
```diff
@@ -0,0 +1,548 @@
...
+This ensures that the height of all the nodes is balanced and hence, an AVL Tree. 
+
+#### Tri-Node Restructuring for Insertion
+
+The procedure for tri-node restructuring for an insertion operation is as follows - 
+
+    1. Perform regular BST insertion.
+    2. Let the node being inserted be known as 'a'. Let 'd' be the first node going up from 'w' towards root that is unbalanced. Let 'c' be a child of 'd' that lies on the path between 'a' and 'd'. Let 'b' be a child of 'c' that comes on the same path. 
+    3. This would result in four possible cases:
+        - Left-left rotation: Where 'c' is the left child of 'd' and 'b' is the left child of 'c'
```
Comment (cyberShaw) [inline articles/introduction-to-avl-trees/index.md:55]: > Left-left rotation: Where 'c' is the left child of 'd' and 'b' is the left child of 'c'

This line would look more readable if you could add bold formatting to these headings.

```md
**Left-Left Rotation**
```

Accordingly, do the same for all rotation types that you've mentioned in the article.

File: articles/introduction-to-avl-trees/index.md [hunk 5]
```diff
@@ -0,0 +1,548 @@
...
+    root = delete_AVL(root, 6);  
+    cout << "The inorder traversal of the AVL tree is: "<<endl;  
+    inOrder(root);
+    cout << endl;
+       
+}
+```
+
+The time complexity of deletion is O(logn) as well. This is due to the self-balancing feature of the tree. 
+
```
Comment (cyberShaw) [inline articles/introduction-to-avl-trees/index.md:546]: A "Further Reading" section with articles related to C++ and AVL Trees would complement your article well.

Example, https://www.section.io/engineering-education/validating-xml-using-dtd/#further-reading

**Review by cyberShaw — COMMENTED** [review]
Great article, @bloodprashure!
It was a good dive into AVL trees.

In the first phase of the review, I've suggested some enhancements mainly related to readability of the article. Please make these changes and we can move on to the tech review.

Also, I noticed some minor grammatical mistakes. Please use a grammar checker like Grammarly to review your article before publishing it. Since the mistakes were very minimal, I fixed them myself in commit:

File: articles/introduction-to-avl-trees/index.md (hunk 1 above)
Comment (bloodprashure) [reply articles/introduction-to-avl-trees/index.md:45]: The images weren't taken from an external source, I made all of them.

File: articles/introduction-to-avl-trees/index.md (hunk 2 above)
Comment (bloodprashure) [reply articles/introduction-to-avl-trees/index.md:8]: Revisions made in the recent commit.

File: articles/introduction-to-avl-trees/index.md (hunk 3 above)
Comment (bloodprashure) [reply articles/introduction-to-avl-trees/index.md]: Revisions made in the recent commit.

File: articles/introduction-to-avl-trees/index.md (hunk 4 above)
Comment (bloodprashure) [reply articles/introduction-to-avl-trees/index.md]: Revisions made in the recent commit.

File: articles/introduction-to-avl-trees/index.md (hunk 5 above)
Comment (bloodprashure) [reply articles/introduction-to-avl-trees/index.md:560]: Revisions made in the recent commit.

File: articles/introduction-to-avl-trees/index.md (hunk 1 above)
Comment (cyberShaw) [reply articles/introduction-to-avl-trees/index.md:45]: Ah, that's great!

**Comment by cyberShaw** [comment]
Thank you for making the suggested fixes, @bloodprashure!

I've added a commit with minor fixes. We can now move on to tech review, where the article would be checked for technical accuracy.

File: articles/introduction-to-avl-trees/index.md [hunk 6]
```diff
@@ -0,0 +1,578 @@
...
+    3. This would result in four possible cases:
+        - **Left-left rotation**: Where 'c' is the left child of 'd' and 'b' is the left child of 'c' 
+        - **Right-right rotation**: Where 'c' is the right child of 'd' and 'b' is the right child of 'c' 
+        - **Left-right rotation**: Where 'c' is the left child of 'd' and 'b' is the right child of 'c' 
+        - **Right-left rotation**: Where 'c' is the right child of 'd' and 'b' is the left child of 'c' 
+    4. After the rotations are performed according to the situation, the node is successfully inserted and the tree is balanced.
+  
+**Left-left Rotation**
+
+<center> <img src="/engineering-education/introduction-to-avl-trees/leftleft_rotation.jpg" alt="Left-Left Rotation"> </center>
```
Comment (cyberShaw) [inline articles/introduction-to-avl-trees/index.md:69]: Aren't the images here redundant and have already been displayed in the previous section?

File: articles/introduction-to-avl-trees/index.md [hunk 7]
```diff
@@ -0,0 +1,578 @@
...
+    // If the first element is being inserted, return it as a node
+    if (node == NULL)
+        return (newNode(val));
+
+    // If the value to be inserted is lesser than the current node, traverse towards the left
+    if (val < node->value)
+       {
+            node->leftChild = insert_AVL(node->leftChild, val);
+            // Balance the nodes after insertion
+            node=balance(node);d
```
Comment (cyberShaw) [inline articles/introduction-to-avl-trees/index.md:236]: > ```node=balance(node);d```

The extra 'd' after the semi-colon results in an error.

```cpp
node=balance(node);
```

File: articles/introduction-to-avl-trees/index.md [hunk 8]
```diff
@@ -0,0 +1,578 @@
...
+
+Since the tree is 'automatically' balanced after every insertion, the complexity of insertion remains at O(logn), unlike a regular BST.
+
+### Deletion from AVL Trees in C++
+
+Deletion follows a very similar algorithm compared to insertion. However, one must keep in mind that different nodes can be deleted i.e. nodes with one child, nodes with two children and nodes with no children. These cases have to be taken into account while balancing the tree.
+
+    1. Perform regular BST deletion.
+    2. Get the balance factor of the current node (height of left subtree - height of right subtree).
+    3. If the balance factor is equal to 2, then the case could be either right-right or right-left rotation. If the balance of the left tree is equal to 1, then right-right rotation, else right-left rotation.
```
Comment (cyberShaw) [inline articles/introduction-to-avl-trees/index.md:305]: There are quite a few errors in the algorithm you've suggested for AVL tree deletion.

Please refer to https://www.geeksforgeeks.org/avl-tree-set-2-deletion/ and check the algorithm once.

**Review by cyberShaw — CHANGES_REQUESTED** [review]
I've done a tech review, @bloodprashure.

I noticed quite a few areas where this article could improve. And there were certain technical inconsistencies in the Deletion implementation. Please check out the comments and fix them. 

Once you're done with those fixes, we can move this article to final review :D

**Comment by cyberShaw** [comment]
@bloodprashure 
These websites could help:
https://www.cs.usfca.edu/~galles/visualization/BST.html
https://www.cs.usfca.edu/~galles/visualization/AVLtree.html

**Comment by bloodprashure** [comment]
My apologies for the technical errors. I shall get them sorted immediately.

**Comment by hectorkambow** [comment]
@cyberShaw 👋 Hi Harsha - just wanted to get a status update on this PR - to see if its ready for a final review?

**Comment by cyberShaw** [comment]
> @cyberShaw 👋 Hi Harsha - just wanted to get a status update on this PR - to see if its ready for a final review?

I'll do the review by today 😄

**Comment by cyberShaw** [comment]
@bloodprashure everything looks good!
There was a small error in the deletion algorithm, so I made the fix in 3a37391
This article is ready for final review 👍

**Comment by bloodprashure** [comment]
> @bloodprashure everything looks good!
> There was a small error in the deletion algorithm, so I made the fix in [3a37391](https://github.com/section-io/engineering-education/commit/3a3739141db77dde1d2ee80633495db91a8895e4)
> This article is ready for final review 👍

Thanks for reviewing, @cyberShaw!

**Comment by cyberShaw** [comment]
I've added the author page for @bloodprashure in db6d34a.

@hectorkambow Please check this and perform the final review 🚀

**Comment by hectorkambow** [comment]
@bloodprashure  
Thank you for your article - we are nearing our final review. 

If you could please replace the image links (from HTML) to .md (markdown) format. 

similar to this. 
[Alt Image Tag](Image-path)

**Comment by bloodprashure** [comment]
> @bloodprashure
> Thank you for your article - we are nearing our final review.
> 
> If you could please replace the image links (from HTML) to .md (markdown) format.
> 
> similar to this.
> [Alt Image Tag](Image-path)

Hey Hector! Thank you for giving me the opportunity!
I have replaced the image links in the recent commit.
Thanks!

**Comment by hectorkambow** [comment]
> > @bloodprashure
> I have replaced the image links in the recent commit.
> Thanks!

Fantastic thank you v much. 

Just to double check these images do not need sources?

**Comment by bloodprashure** [comment]
> > > @bloodprashure
> > > I have replaced the image links in the recent commit.
> > > Thanks!
> 
> Fantastic thank you v much.
> 
> Just to double check these images do not need sources?

I made them myself, so no, they will not be needing sources :)

**Pull request merged**